toolchain/lexer/tokenized_buffer.cpp

```diff
@@ -48,6 +52,164 @@ auto VariantMatch(V&& v, Fs&&... fs) -> decltype(auto) {
+  // identifier (or keyword) characters. This is used in the generic,
+  // non-vectorized fallback code to scan for length of an identifier.
+  constexpr std::array<bool, 256> IsIdentifierByteTable = []() constexpr {
+    std::array<bool, 256> table = {};
+    for (char c = '0'; c <= '9'; ++c) {
+      table[c] = true;
+    }
+    for (char c = 'A'; c <= 'Z'; ++c) {
+      table[c] = true;
+    }
+    for (char c = 'a'; c <= 'z'; ++c) {
+      table[c] = true;
+    }
+    table['_'] = true;
+    return table;
+  }();
+
+#if __x86_64__
+  // This code uses a scheme derived from the techniques in Geoff Langdale and
+  // Daniel Lemire's work on parsing JSON[1]. Specifically, that paper outlines
+  // a technique of using two 4-bit indexed in-register look-up tables (LUTs) to
+  // classify bytes in a branchless SIMD code sequence.
+  //
+  // [1]: https://arxiv.org/pdf/1902.08318.pdf
+  //
+  // The goal is to get a bit mask classifying different sets of bytes. For each
+  // input byte, we want the mask bits to be set with the following logic
+  // derived by inspecting the high nibble and low nibble of the input:
+  // bit0 = 1 for `_` needs 1 for high `0x5` and low `0xF`
+  // bit1 = 1 for `0-9` needs 1 for high `0x3` and low `0x0` - `0x9`
+  // bit2 = 1 for `A-O` needs 1 for high `0x4` and low `0x1` - `0xF`
+  // bit3 = 1 for `P-Z` needs 1 for high `0x5` and low `0x0` - `0xA`
+  // bit4 = 1 for `a-o` needs 1 for high `0x6` and low `0x1` - `0xF`
+  // bit5 = 1 for `p-z` needs 1 for high `0x7` and low `0x0` - `0xA`
+  // bit6 = unused
+  // bit7 = 1 for possible non-ASCII character in UTF-8, for all high nibbles
+  // with the high bit set and all low-nibbles.
+  //
+  // No bits set means definitively non-ID ASCII character.
+  //
+  // High bit set means contains the start of a non-ASCII character, need to
+  // fall back to a UTF-8 decoding path (not yet fully implemented).
+  //
+  // bit6 remains unused if we need to classify more characters.
+  const auto high_lut = _mm_setr_epi8(
+      /*0x0:*/ 0b0000'0000,
+      /*0x1:*/ 0b0000'0000,
+      /*0x2:*/ 0b0000'0000,
+      /*0x3:*/ 0b0000'0010,
+      /*0x4:*/ 0b0000'0100,
+      /*0x5:*/ 0b0000'1001,
+      /*0x6:*/ 0b0001'0000,
+      /*0x7:*/ 0b0010'0000,
+      /*0x8:*/ 0b1000'0000,
+      /*0x9:*/ 0b1000'0000,
+      /*0xA:*/ 0b1000'0000,
+      /*0xB:*/ 0b1000'0000,
+      /*0xC:*/ 0b1000'0000,
+      /*0xD:*/ 0b1000'0000,
+      /*0xE:*/ 0b1000'0000,
+      /*0xF:*/ 0b1000'0000);
+  const auto low_lut = _mm_setr_epi8(
+      /*0x0:*/ 0b1010'1010,
+      /*0x1:*/ 0b1011'1110,
+      /*0x2:*/ 0b1011'1110,
+      /*0x3:*/ 0b1011'1110,
+      /*0x4:*/ 0b1011'1110,
+      /*0x5:*/ 0b1011'1110,
+      /*0x6:*/ 0b1011'1110,
+      /*0x7:*/ 0b1011'1110,
+      /*0x8:*/ 0b1011'1110,
+      /*0x9:*/ 0b1011'1110,
+      /*0xA:*/ 0b1011'1100,
+      /*0xB:*/ 0b1001'0100,
+      /*0xC:*/ 0b1001'0100,
+      /*0xD:*/ 0b1001'0100,
+      /*0xE:*/ 0b1001'0100,
+      /*0xF:*/ 0b1001'0101);
```

Comment:
Maybe instead put this in that kind of form, then? It looks like parameters are named b0 through b15:

```
_mm_setr_epi8(char __b0, char __b1, char __b2, char __b3, char __b4, char __b5,
              char __b6, char __b7, char __b8, char __b9, char __b10,
              char __b11, char __b12, char __b13, char __b14, char __b15) {
```

So wouldn't style be `/*__b0=*/`, etc?

Note, if you don't want to do this, it's probably worth thinking about what it means for Carbon if a function in a library that we don't own has parameter names you don't find useful.

Reply:
I'll try this and see if I can get the parameter name approach to work well. Even though they're not the *most* readable, this would have actually caught a bug I wrote the first time (I used `_mm_set_epi8` instead).

That said, not sure it makes sense to generalize too much to Carbon here -- I feel like there are a bunch of overlapping challenges from C++ here and we might better off just working with Carbon code that ends up with bit tables and similar constructs in its API design, and see what the best tools are or what changes would give better results there.